src/mon/OSDMonitor.cc

```diff
@@ -1789,13 +1789,9 @@ bool OSDMonitor::prepare_failure(MonOpRequestRef op)
 		       << m->get_orig_source_inst() << "\n";
     if (failure_info.count(target_osd)) {
       failure_info_t& fi = failure_info[target_osd];
-      list<MonOpRequestRef> ls;
-      fi.take_report_messages(ls);
-      fi.cancel_report(reporter);
-      while (!ls.empty()) {
-        if (ls.front())
-          mon->no_reply(ls.front());
-	ls.pop_front();
+      MonOpRequestRef record_op = fi.cancel_report(reporter);
```

Comment:
note to myself and posterity:

if osd.42 is reported to as "failed" by its peer. and later the peer receives a ping_reply  from osd.42 somehow, so it sends a failure message to monitor to cancel the previous failure report. as a side-effect, on the monitor handling this failure report, all report messages from other OSD peers were also erased from the `failure_info`. but their osd ids and `failed_since` are kept around in the `failure_info`. and they will be taken into consideration in `OSDMonitor::check_failure()` even the report messages are reset.

if osd.42 is finally marked down, all the osds not reverting their failure report will **not** be updated with the latest osdmap. but we could.

the related change was introduced by ad12b0d61b8c4565ee8a5f352689b9121da2bcd2, so we don't leak the failure messages before canceling the report. but `failure_report` and `failure_reporter_t` are able to take care of their life cycles just fine without the fix.